app/scripts/controller-init/types.ts

```diff
@@ -0,0 +1,168 @@
+import { Provider } from '@metamask/network-controller';
+import {
+  ActionConstraint,
+  ControllerMessenger,
+  EventConstraint,
+  RestrictedControllerMessenger,
+} from '@metamask/base-controller';
+import { TransactionMetricsRequest } from '../lib/transaction/metrics';
+import { Controller, ControllerFlatState } from './controller-list';
+
+/** The supported controller names. */
+export type ControllerName = Controller['name'];
+
+/** All controller types by name. */
+export type ControllerByName = {
+  [name in ControllerName]: Controller & { name: name };
+};
+
+/**
+ * Persisted state for all controllers.
+ * e.g. `{ TransactionController: { transactions: [] } }`.
+ */
+export type ControllerPersistedState = {
+  [name in ControllerName]: ControllerByName[name]['state'];
+};
+
+/** Generic controller messenger using base template types. */
+export type BaseControllerMessenger = ControllerMessenger<
+  ActionConstraint,
+  EventConstraint
+>;
+
+/** Generic restricted controller messenger using base template types. */
+export type BaseRestrictedControllerMessenger = RestrictedControllerMessenger<
+  string,
+  ActionConstraint,
+  EventConstraint,
+  string,
+  string
+>;
+
+/**
+ * Request to initialize and return a controller instance.
+ * Includes standard data and methods not coupled to any specific controller.
+ */
+export type ControllerInitRequest<
+  ControllerMessengerType extends void | BaseRestrictedControllerMessenger = void,
+  InitMessengerType extends void | BaseRestrictedControllerMessenger = void,
+> = {
+  /**
+   * Required controller messenger instance.
+   * Generated using the callback specified in `getControllerMessengerCallback`.
+   */
+  controllerMessenger: ControllerMessengerType extends BaseRestrictedControllerMessenger
```

Comment:
Not sure I follow what this conditional type was meant to achieve. The `ControllerMessengerType` type is optional, but that inevitably results in `never` here. It's both optional and not optional.

I suppose this was just for ease of testing? If so, maybe we can deal with that another way instead, and keep `ControllerMessengerType` as mandatory.

Reply:
Though, at least for the init messenger, it does make sense for it to be optional. That wouldn't always be needed.

In that case, perhaps we can make the type less of a contradiction by only including the `initMessenger` property if the `InitMessengerType` is given.

Reply:
Done, left the init messenger as optional but switched syntax so the property no longer exists rather than having `never` type.